primitives/src/merkle_tree/merkle_tree_traits.rs

```diff
@@ -0,0 +1,158 @@
+
+use crate::errors::PrimitivesError;
+use ark_ff::Field;
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+use ark_std::{borrow::Borrow, fmt::Debug, string::ToString};
+use typenum::Unsigned;
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug, Hash)]
+/// The result of querying at an index in the tree
+pub enum LookupResult<F, P> {
+    /// The value at the given index, and a proof of validity
+    Ok(F, P),
+    /// The index is valid but we do not have the leaf in memory
+    NotInMemory,
+    /// The index is outside the occupied range in the tree
+    EmptyLeaf,
+}
+
+impl<F, P> LookupResult<F, P> {
+    /// Assert the lookup result is Ok.
+    pub fn expect_ok(self) -> Result<(F, P), PrimitivesError> {
+        match self {
+            LookupResult::Ok(x, proof) => Ok((x, proof)),
+            LookupResult::NotInMemory => Err(PrimitivesError::InternalError(
+                "Expected Ok, found NotInMemory".to_string(),
+            )),
+            LookupResult::EmptyLeaf => Err(PrimitivesError::InternalError(
+                "Expected Ok, found EmptyLeaf".to_string(),
+            )),
+        }
+    }
+}
+
+/// Merkle tree element type
+pub trait ElementType<F: Field>:
+    Default
+    + Ord
+    + Clone
+    + Copy
+    + Debug
+    + CanonicalSerialize
+    + CanonicalDeserialize
+    + Eq
+    + PartialEq
+    + AsRef<[F]>
+{
+    /// Convert the element into a slice of field elements
+    /// TODO: is it the same as `AsRef<[F]>`?
+    fn as_slice_ref(&self) -> &[F];
+}
+
+/// Merkle tree hash function
+pub trait Hasher<F: Field> {
+    /// Digest a list of values
+    fn digest(data: &[F]) -> F;
+}
+
+/// Basic functionalities for a merkle tree implementation
+pub trait MerkleTree<F: Field>: Sized {
+    /// Merkle tree element type
+    type ElementType: ElementType<F>;
+    /// Hash algorithm used in merkle tree
+    type Hasher: Hasher<F>;
+    /// Leaf arity
+    type LeafArity: Unsigned;
+    /// Non-leaf arity
+    type TreeArity: Unsigned;
+    /// Exsistential proof
+    type Proof;
+    /// Batch proof
+    type BatchProof;
+
+    /// Construct a new merkle tree with given height from a data slice
+    fn build(
+        height: usize,
+        data: impl Iterator<Item = Self::ElementType>,
+    ) -> Result<Self, PrimitivesError>;
+
+    /// Return the height of this merkle tree
+    fn height(&self) -> usize;
+    /// Return the maximum allowed number leaves
+    fn capacity(&self) -> u64;
+    /// Return the current number of leaves
+    fn num_leaves(&self) -> u64;
+
+    /// Return the current root value
+    fn value(&self) -> F;
+
+    /// Returns the leaf value given a position
+    /// * `pos` - zero-based index of the leaf in the tree
+    /// * `returns` - Leaf value at the position. LookupResult::EmptyLeaf if the
+    ///   leaf position is empty or invalid, LookupResult::NotInMemory if the
+    ///   leaf position has been forgotten.
+    fn lookup(&self, pos: u64) -> LookupResult<(), Self::Proof>;
```

Comment:
Why return `()` and not `Self::ElementType`?

Reply:
I'm not sure whether it's useful to return the element. But let me add this.